File: CCC 2007/s5 - bowling for numbers.cpp
#include <iostream>
#include <stdio.h>
#include <string.h>
using namespace std;
int dp[501][60001], scores[60001], sums[60001], pins, balls, width;
int sum(int left, int right){
    int s = 0;
    for(int i = left; i < right; i++){
        s += scores[i];
    }
    return s;
}
int max(int a, int b){
    return a > b ? a : b;
}
int main(){
    int cases;
    scanf("%d", &cases);
    while(cases--){
        scanf("%d%d%d", &pins, &balls, &width);
        int setW = (pins + width + 1);
        memset(dp[0], 0, sizeof(dp[0][0]) * setW);
        memset(scores, 0, sizeof(scores[0]) * setW);
        for(int i = width; i < width+pins; i++){
            scanf("%d", &scores[i]);
            sums[i] = sum(i-width+1, i+1);
        }
        int m = 0;
        for(int row = 1; row < balls+1; row++){
            for(int col = width; col < width + pins; col++){
                dp[row][col] = max(sums[col] + dp[row-1][col-width], dp[row][col-1]);
                if(row == balls){
                    m = max(m, dp[row][col]);
                }
            }
        }
        printf("%d\n", m);
    }
}

File: CCC 2007/s5 - bowling for numbers.py
# See C++ Version for details on how to solve

cases = int(raw_input())

dp = [[0] * 60001 for j in range(0, 501)]
scores = [0] * 60001
sums = [0] * 60001

for iasd in range(0, cases):
	args = map(int, raw_input().split())
	pins = args[0]
	balls = args[1]
	width = args[2]
	arrW = (pins + width + 2)
	for i in range(0, arrW):
		dp[0][i] = 0
		scores[i] = 0
		sums[i] = 0
	for i in range(width, pins + width):
		scores[i] = int(raw_input())
		sums[i] = sum(scores[i-width+1:i+1])

	for row in range(1, balls+1):
		for pin in range(width, pins + width):
			dp[row][pin] = max(sums[pin] + dp[row-1][pin-width], dp[row][pin-1])
			
	print max(dp[balls])